src/eos-updater-poll-common.c

```diff
@@ -312,6 +310,118 @@ get_booted_refspec (gchar               **booted_refspec,
   return TRUE;
 }
 
+static gboolean
+get_refspec_to_upgrade_on_from_deployment (OstreeSysroot     *sysroot,
+                                           OstreeDeployment  *booted_deployment,
+                                           gchar            **out_refspec_to_upgrade_from_deployment,
+                                           GError           **error)
+{
+  const gchar *checksum = ostree_deployment_get_csum (booted_deployment);
+  g_autoptr(GVariant) commit = NULL;
+  g_autoptr(GVariant) metadata = NULL;
+  g_autoptr(GVariant) ref_for_deployment_variant = NULL;
+  g_autofree gchar *refspec_for_deployment = NULL;
+  OstreeRepo *repo = NULL;
+
+  g_return_val_if_fail (out_refspec_to_upgrade_from_deployment != NULL, FALSE);
+
+  if (!ostree_sysroot_get_repo (sysroot, &repo, NULL, error))
+   return FALSE;
+
+  if (!ostree_repo_load_variant (repo,
+                                 OSTREE_OBJECT_TYPE_COMMIT,
+                                 checksum,
+                                 &commit,
+                                 error))
+    return FALSE;
+
+  /* Look up the checkpoint target to see if there is one on this commit. */
+  metadata = g_variant_get_child_value (commit, 0);
+  ref_for_deployment_variant = g_variant_lookup_value (metadata,
+                                                       "eos.checkpoint-target",
+                                                       G_VARIANT_TYPE_STRING);
```

Comment:
Since you’re going to have to rebase and squash everything here anyway, would it be possible to split the checkpoint stuff (the changes here in `poll-common.c` and the new tests for it) into a new PR, which I can review and merge separately (probably before this PR gets merged). That should help to unblock things — generally splitting large branches like this up into easily-mergeable groundwork PRs, followed by a larger PR for the new feature where its interface with the existing code is minimal and easily verifiable to be correct, is good. The checkpoint stuff counts as groundwork, as does splitting `test-common` out.

wrt rebasing everything else once you’ve split those out, I think it might be easiest if you just squashed everything together, then split it out again a bit as necessary (using `git reset HEAD~; while true; do git add -p; git commit; done`). It seems like a lot of the commits in the branch are thinly-veiled fixups, and trying to disentangle them all and squash them in to the right commits individually is going to be too hard. Given that you’re ultimately not going to end up with many commits on the branch (since the feature is pretty monolithic once the groundwork has been split out), it’s probably easiest to start from a single monolithic commit and break it down a bit.

I hope that makes sense. If not, poke me and we can discuss it. I should be able to find time to do the rebase/rearrangement of the branch myself if you think that would help or be instructive. I don’t want to step on your toes with it if you have a plan already though. :-)